File: DMOJ/ECOO/ChrisCandy.cpp
// ECOO '21 P4 - Chris' Candy

#include<cstdio>
#include<unordered_map>
#include<cmath>
typedef long long ll;

int ttl = 0;
ll k, tempk;
std::unordered_map<ll, int> facs;
int main(){
    scanf("%lld", &k);
    k++;
    tempk = k;
    while (!(k&1)){
        k >>= 1;
        facs[2]++;
        ttl++;
    }
    for (int i=3; i<=sqrt(k); i++){
        while (!(k%i)){
            k /= i;
            facs[i]++;
            ttl += i-1;
        }
    }

    if (k != 1){
        if (k > 100000){
            printf("Sad Chris\n");
            return 0;
        }
        facs[k] = 1;
        ttl += k-1;
    }

    if (ttl > 100000){
        printf("Sad Chris\n");
        return 0;
    }

    printf("%d\n", ttl);
    int ind = 1;
    for (auto p: facs){
        for (int i=0; i<p.second; i++){
            for (int j=1; j<p.first; j++){
                printf("%d ", ind);
            }
            ind++;
        }
    }

    return 0;
}

File: DMOJ/ECOO/ChrisCandy.py
# ECOO '21 P4 - Chris' Candy

from math import sqrt


def main():
    k = int(input()) + 1
    cop = k

    # prime factorize (input+1)
    fac = {}
    if not k&1:
        fac[2] = 0
    while not k&1:
        fac[2] += 1
        k >>= 1

    for i in range(3, int(sqrt(k))+1):
        if not k%i:
            fac[i] = 0
        while not k%i:
            fac[i] += 1
            k //= i

    if k != 1:
        fac[k] = 1

    if sum(fac.values()) == 1:
        # k is prime
        if cop <= 100000:
            print(cop-1)
            for i in range(cop-1):
                print(1, end=' ')
        else:
            print("Sad Chris")
    else:
        ind = 1
        ans = 0
        output = {}
        for val, cnt in fac.items():
            for _ in range(cnt):
                ans += val-1
                output[ind] = val-1
                ind += 1
            if ans > 100000:
                print("Sad Chris")
                return
        print(ans)
        for val, cnt in output.items():
            for _ in range(cnt):
                print(val, end=' ')

main()
